File: app/src/main/java/net/rmoreno/weatherapp/ui/MainActivity.java
package net.rmoreno.weatherapp.ui;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.provider.Settings;
import android.support.v7.widget.CardView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import net.rmoreno.weatherapp.MyLocationListener;
import net.rmoreno.weatherapp.R;
import net.rmoreno.weatherapp.WeatherInteractor;
import net.rmoreno.weatherapp.adapters.DailyAdapter;
import net.rmoreno.weatherapp.models.CurrentWeather;
import net.rmoreno.weatherapp.models.DailyWeather;
import net.rmoreno.weatherapp.presenters.WeatherPresenter;
import net.rmoreno.weatherapp.presenters.WeatherPresenterImpl;
import net.rmoreno.weatherapp.repositories.WeatherRepository;

import java.util.ArrayList;


public class MainActivity extends Activity implements WeatherView {

    String ACTIVITY = "MAIN ACTIVITY";
    String SHARED_PREFERENCES = "MyPrefs";
    CardView cardView;

    RecyclerView mRecyclerView;

    TextView temperature;
    TextView summary;
    TextView time;
    TextView precipitation;
    TextView wind;
    TextView feelsLike;

    ImageView icon;
    public WeatherPresenter weatherPresenter;

    int REQUEST_CODE = 100;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //set sharedPreferences preferences to detect if user has a 'sweater weather'
        SharedPreferences sharedPreferences = getSharedPreferences(SHARED_PREFERENCES, Context.MODE_PRIVATE);


        weatherPresenter = new WeatherPresenterImpl(
                this,
                new WeatherInteractor(
                        new WeatherRepository(sharedPreferences)
                )
        );

        if(weatherPresenter.firstTime()) {
            Intent intent = new Intent(MainActivity.this, IntroActivity.class);
            startActivity(intent);
            Toast.makeText(MainActivity.this, "sweater is 0", Toast.LENGTH_SHORT).show();
        }

        mRecyclerView = (RecyclerView) findViewById(R.id.recycler_view);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        cardView = (CardView) findViewById(R.id.currentData);

        temperature = (TextView) findViewById(R.id.temperature);
        summary = (TextView) findViewById(R.id.summary_text);
        time = (TextView) findViewById(R.id.time);
        precipitation = (TextView) findViewById(R.id.precipitation);
        feelsLike = (TextView) findViewById(R.id.feels);
        wind = (TextView) findViewById(R.id.wind);

        icon = (ImageView) findViewById(R.id.current_icon);

        if(isNetworkAvailible()) {
            getLocation();
        }
    }

    @Override
    public void onResume(){
        super.onResume();
        weatherPresenter.resume();
        getLocation();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

    }

    private boolean isNetworkAvailible() {
        ConnectivityManager manager = (ConnectivityManager)
                getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = manager.getActiveNetworkInfo();
        boolean isAvailable = false;
        if (networkInfo != null && networkInfo.isConnected()){
            isAvailable = true;
        }
        return isAvailable;
    }

    public boolean isLocationEnabled(LocationManager lm){
        boolean gps_enabled = false;
        boolean network_enabled = false;

        try {
            gps_enabled = lm.isProviderEnabled(LocationManager.GPS_PROVIDER);
        } catch(Exception ex) {}

        try {
            network_enabled = lm.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
        } catch(Exception ex) {}

        if(!gps_enabled && !network_enabled) {
            return false;
        } else{
            return true;
        }
    }

    @Override
    public void displayDailyWeather(ArrayList<DailyWeather> dailyWeather, int sweaterTemp) {
        DailyAdapter adapter = new DailyAdapter(MainActivity.this, dailyWeather, sweaterTemp);
        mRecyclerView.setAdapter(adapter);
    }

    @Override
    public void displayCurrentWeather(CurrentWeather currentWeather) {
        temperature.setText(currentWeather.getTemp() + "°");
        time.setText("At " + currentWeather.getFormatedTime());
        icon.setImageResource(currentWeather.getIconId());
        summary.setText(currentWeather.getSummary());
        precipitation.setText(String.valueOf(currentWeather.getPrecip())+"%");
        feelsLike.setText(String.valueOf(currentWeather.getFeels())+ "°");
        wind.setText(String.valueOf(currentWeather.getWind()) + "mph");
    }

    public void buildDialog(Context context){
        AlertDialog.Builder dialog = new AlertDialog.Builder(context);
        dialog.setMessage("GPS network is not enabled");
        dialog.setPositiveButton(context.getResources().getString(R.string.open_location_settings), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface paramDialogInterface, int paramInt) {
                Intent myIntent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                startActivityForResult(myIntent, REQUEST_CODE);
            }
        });

        dialog.setNegativeButton(context.getString(R.string.Cancel), new DialogInterface.OnClickListener() {

            @Override
            public void onClick(DialogInterface paramDialogInterface, int paramInt) {

            }
        });
        dialog.show();
    }

    public void getLocation() {
        LocationManager lm = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
        LocationListener locationListener = new MyLocationListener();

        lm.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0,0, locationListener);
        if(isLocationEnabled(lm)) {
            Location location = lm.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
            Log.d("LOCATION",String.valueOf(location.getLatitude()));
            weatherPresenter.getCurrentWeather(location.getLatitude(), location.getLongitude());
            weatherPresenter.getDailyWeather(location.getLatitude(), location.getLongitude());

        } else {
            buildDialog(MainActivity.this);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_settings) {
            return true;
        } else if(id == R.id.set_sweather) {
            Intent intent = new Intent(MainActivity.this, SettingsActivity.class);
            startActivity(intent);
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/net/rmoreno/weatherapp/ui/MainActivity.kt
package net.rmoreno.weatherapp.ui

import android.Manifest
import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.activity_main.*
import net.rmoreno.weatherapp.*
import net.rmoreno.weatherapp.adapters.DailyAdapter
import net.rmoreno.weatherapp.models.Currently
import net.rmoreno.weatherapp.models.DailyDetail
import net.rmoreno.weatherapp.models.ForecastResponse
import net.rmoreno.weatherapp.presenters.BasePresenter
import net.rmoreno.weatherapp.presenters.WeatherPresenter
import net.rmoreno.weatherapp.repositories.WeatherRepository

class MainActivity : Activity(), WeatherView {

    private var SHARED_PREFERENCES = "MyPrefs"

    private lateinit var weatherPresenter: BasePresenter<ForecastResponse, WeatherView>

    private var REQUEST_CODE = 100

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            checkPermission()
        }
        weather_loading_bar.visibility = View.VISIBLE
        weather_loading_bar.bringToFront()
    }

    public override fun onResume() {
        super.onResume()

        val sharedPreferences = getSharedPreferences(SHARED_PREFERENCES, Context.MODE_PRIVATE)
        val locationSensor = LocationSensor(this@MainActivity)
        val weatherRepository = WeatherRepository(sharedPreferences)
        val weatherInteractor = WeatherInteractor(weatherRepository, locationSensor)
        weatherPresenter = WeatherPresenter(weatherInteractor)
        weatherPresenter.bindView(this@MainActivity)
    }

    public override fun onPause() {
        super.onPause()
        weatherPresenter.unbindView()
    }

    override fun goToIntroActivity() {
        val intent = Intent(this@MainActivity, IntroActivity::class.java)
        startActivity(intent)
    }

    override fun displayDailyWeather(dailyWeather: List<DailyDetail>, sweaterTemp: Int, timezone: String) {
        val dailyAdapter = DailyAdapter(sweaterTemp, timezone)
        recycler_view.apply {
            adapter = dailyAdapter
            layoutManager = LinearLayoutManager(this@MainActivity)
        }
        dailyAdapter.addData(dailyWeather.toMutableList())
    }

    override fun displayCurrentWeather(currentWeather: Currently, timezone: String) {
//        val num = Math.round(currentWeather.precipProbability)
//        precipitation.text = getString(R.string.precip_probability, 1)
        temperature.text = getString(R.string.degrees, Math.round(currentWeather.temperature))
        //TODO: make this shake when we can't update the weather due to network errors or lack of location
        time.text = getString(R.string.time_at, TimeUtil.formatTime(currentWeather.time, timezone))
        current_icon.setImageResource(IconUtil.getIcon(currentWeather.icon))
        summary_text.text = currentWeather.summary
        feels.text = getString(R.string.degrees, Math.round(currentWeather.apparentTemperature))
        wind.text = getString(R.string.wind_speed, Math.round(currentWeather.windSpeed))
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun setLoading(isLoading: Boolean) {
        weather_loading_bar.visibility = if (isLoading) View.VISIBLE else View.GONE;
    }

    override fun displayNetworkError() {
        val context: Context = this@MainActivity

        val dialog = AlertDialog.Builder(context)
        dialog.setMessage(getString(R.string.network_not_available))
        dialog.setPositiveButton(context.resources.getString(R.string.open_location_settings)) { _, _ ->
            val myIntent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
            startActivityForResult(myIntent, REQUEST_CODE)
        }

        dialog.setNegativeButton(context.getString(R.string.Cancel)) { _, _ -> }
        dialog.show()
    }

    private fun checkPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED
        ) {

            ActivityCompat.requestPermissions(this,
                    listOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION).toTypedArray(),
                    123);
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId

        when (id) {
            R.id.action_settings -> {
                //TODO retrieve location
                return true
            }

            R.id.set_sweather -> {
                val intent = Intent(this@MainActivity, SettingsActivity::class.java)
                startActivity(intent)
            }

            R.id.refresh -> {
                //TODO implement refresh action
            }
        }
        return super.onOptionsItemSelected(item)
    }
}